flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/taskmanager/ThreadDumpInfo.java

```diff
@@ -106,5 +112,87 @@ public boolean equals(Object o) {
+                            "\""
+                                    + threadInfo.getThreadName()
+                                    + "\""
+                                    + " Id="
+                                    + threadInfo.getThreadId()
+                                    + " "
+                                    + threadInfo.getThreadState());
+            if (threadInfo.getLockName() != null) {
+                sb.append(" on " + threadInfo.getLockName());
+            }
+            if (threadInfo.getLockOwnerName() != null) {
+                sb.append(
+                        " owned by \""
+                                + threadInfo.getLockOwnerName()
+                                + "\" Id="
+                                + threadInfo.getLockOwnerId());
+            }
+            if (threadInfo.isSuspended()) {
+                sb.append(" (suspended)");
+            }
+            if (threadInfo.isInNative()) {
+                sb.append(" (in native)");
+            }
+            sb.append('\n');
+            int i = 0;
+            StackTraceElement[] stackTraceElements = threadInfo.getStackTrace();
+            for (; i < stackTraceElements.length && i < maxDepth; i++) {
+                StackTraceElement ste = stackTraceElements[i];
+                sb.append("\tat " + ste.toString());
+                sb.append('\n');
+                if (i == 0 && threadInfo.getLockInfo() != null) {
+                    Thread.State ts = threadInfo.getThreadState();
+                    switch (ts) {
+                        case BLOCKED:
+                            sb.append("\t-  blocked on " + threadInfo.getLockInfo());
+                            sb.append('\n');
+                            break;
+                        case WAITING:
+                        case TIMED_WAITING:
+                            sb.append("\t-  waiting on " + threadInfo.getLockInfo());
+                            sb.append('\n');
+                            break;
+                        default:
+                    }
+                }
+
+                for (MonitorInfo mi : threadInfo.getLockedMonitors()) {
+                    if (mi.getLockedStackDepth() == i) {
+                        sb.append("\t-  locked " + mi);
+                        sb.append('\n');
+                    }
+                }
+            }
+            if (i < threadInfo.getStackTrace().length) {
+                sb.append("\t...");
+                sb.append('\n');
+            }
+
+            LockInfo[] locks = threadInfo.getLockedSynchronizers();
+            if (locks.length > 0) {
+                sb.append("\n\tNumber of locked synchronizers = " + locks.length);
+                sb.append('\n');
+                for (LockInfo li : locks) {
+                    sb.append("\t- " + li);
+                    sb.append('\n');
+                }
+            }
+            sb.append('\n');
+            return sb.toString();
```

Comment:
Are we sure that this works across Java versions?

Reply:
It works well with Java8 and Java11. And other java versions looks will not be supported on latest Flink release, just refer to https://docs.ververica.com/installation/versions/stream.html#java-compatibility

Reply:
I think the community is actively working on adding support for Java 17. See https://issues.apache.org/jira/browse/FLINK-15736. Maybe we can check whether your changes work with Java 17.

Reply:
It looks good when setting the Java17 in my IDEA.

Reply:
This is nice. Shall we also add a test that ensures that `stringifyThreadInfo` produces the same result as `ThreadInfo.toString` when the depth is set to 8? That way we get notified whenever something changes on the Java side.

Reply:
The `stringifyThreadInfo` lacks partial params due to no public method for obtaining parameters in JDK `ThreadInfo`.
JDK default implemetation please refer to: https://github.com/openjdk/jdk/blob/master/src/java.management/share/classes/java/lang/management/ThreadInfo.java#L597